core/src/main/scala/org/apache/spark/deploy/security/AbstractCredentialRenewer.scala

```diff
@@ -0,0 +1,224 @@
+ *
+ * This class will manage the kerberos login, by renewing the TGT when needed. Because the UGI API
+ * does not expose the TTL of the TGT, a configuration controls how often to check that a relogin is
+ * necessary. This is done reasonably often since the check is a no-op when the relogin is not yet
+ * needed. The check period can be overridden in the configuration.
+ *
+ * New delegation tokens are created once 75% of the renewal interval of the original tokens has
+ * elapsed. The new tokens are sent to the Spark driver endpoint once it's registered with the AM.
+ * The driver is tasked with distributing the tokens to other processes that might need them.
+ *
+ * This class can also be used when without a principal and keytab, in which case token renewal will
+ * not be available. It provides a different API in that case (see `createAndUpdateTokens()`), which
+ * automates the distribution of tokens to the different processes in the Spark app.
+ */
+private[spark] abstract class AbstractCredentialRenewer(
+    protected val sparkConf: SparkConf,
+    protected val hadoopConf: Configuration) extends Logging {
+
+  private val principal = sparkConf.get(PRINCIPAL).orNull
+  private val keytab = sparkConf.get(KEYTAB).orNull
+
+  if (principal != null) {
+    require(keytab != null, "Kerberos principal specified without a keytab.")
+    require(new File(keytab).isFile(), s"Cannot find keytab at $keytab.")
+  }
+
+  private val renewalExecutor: ScheduledExecutorService =
+    if (principal != null) {
+      ThreadUtils.newDaemonSingleThreadScheduledExecutor("Credential Renewal Thread")
+    } else {
+      null
+    }
+
+  private val driverRef = new AtomicReference[RpcEndpointRef]()
+
+  protected def setDriverRef(ref: RpcEndpointRef): Unit = {
+    driverRef.set(ref)
+  }
+
+  protected def renewalEnabled: Boolean = principal != null
+
+  /**
+   * Start the token renewer. Upon start, if a principal has been configured, the renewer will:
+   *
+   * - log in the configured principal, and set up a task to keep that user's ticket renewed
+   * - obtain delegation tokens from all available providers
+   * - schedule a periodic task to update the tokens when needed.
+   *
+   * @return The newly logged in user.
+   */
+  def start(): UserGroupInformation = {
+    require(renewalEnabled, "Token renewal is disabled.")
+
+    val originalCreds = UserGroupInformation.getCurrentUser().getCredentials()
+    val ugi = doLogin()
+
+    val tgtRenewalTask = new Runnable() {
+      override def run(): Unit = {
+        ugi.checkTGTAndReloginFromKeytab()
+      }
+    }
+    val tgtRenewalPeriod = sparkConf.get(KERBEROS_RELOGIN_PERIOD)
+    renewalExecutor.scheduleAtFixedRate(tgtRenewalTask, tgtRenewalPeriod, tgtRenewalPeriod,
+      TimeUnit.SECONDS)
+
+    val creds = obtainTokensAndScheduleRenewal(ugi)
+    ugi.addCredentials(creds)
+
+    val driver = driverRef.get()
+    if (driver != null) {
```

Comment:
Should a `require(driver.isDefined)` be thrown here to ensure that the driver sends the updates?

Reply:
YARN starts the token renewer before there is a driver reference available; the AM gets that later (when it connects back to the driver in client mode, or when the user code is started in cluster mode).

I've been trying to come up with a better API here that would avoid these checks, but haven't figure it out yet.

Basically for the YARN case you would need:

- an initial call to login and fetch delegation tokens, before the driver is even known (the fix for SPARK-23361)
- a call to start the renewal task, which needs information retrieved in the above call (the renewal interval)

So breaking it into two calls would need some state to be kept between the calls. Not sure whether that would be much cleaner than what's here now.

Trying to follow the Mesos way (where the driver itself manages the renewer, instead of the AM code in YARN) would probably not work in YARN, because the AM tries to connect to HDFS (for distributed cache functionality) before the driver is running, so it needs those tokens. It might be possible to rework some of the YARN backend internals to avoid that, I think, but it needs more thought and would be better left as a separate / future enhancement. 

At some point I'd also like to change the client mode on YARN to be more like the Mesos case (where the driver manages the login, not the AM), maybe all this could be cleaned up at that time.

Reply:
What about extending this start() method with a boolean flag sendingInitialToken so it would be clear why this functionality is here and also this would not be controlled by the having driver registered beforehand (would be easy to see for Mesos this is a must have so the require can be used too in the if body).

Reply:
> What about extending this start() method with a boolean flag sendingInitialToken

That is just as awkward, because then you'd have to set the driver ref before calling start with that flag set to true.

The API is just going to be a little awkward until we refactor the way the YARN code deals with tokens. The goal here is not to fix the API, but to share all the rest of the logic.